### ISSUE TITLE
Question: Have you been able to run PERL code in Lambda?

### ISSUE THREAD

**Issue by kittrCZ** [pplu/perl-aws-lambda#1]
Hi,

I was interested in knowing how successfully have you been able to run Perl in AWS Lambda environment? We are looking to host http://www.sno.phy.queensu.ca/~phil/exiftool/ in Lambda, but it's Perl-based library.

Have you found some difficulties you would share with me?

Best,
Tomas

**Comment by pplu** [comment]
Hi! Yes! I was able to run Perl on Lambda. I maintained correspondence some months with a person that was interested in the same. My intention was to pusblish it in some way, but never got down to it. So here is a quick dump of our mails:

```
The perl-aws-lambda is a proof of concept on how to execute Perl from withinside Lambda (and it's possible!)
 
It results that the AMI that Lambda executes on has Perl installed (it's a public image). You can upload to Lambda whatever you want (any binary) that will execute on that AMI. You can invoke that binary (or preinstalled binaries) from your Lambda function (yay!)
 
For Perl I see various degrees of things you can do:
 
All is based on what you put inside the zip you upload to Lambda
 
1- execute a single Perl script
 
That's trivial. The Lambda function invokes the single Perl script and passes a serialized version of it's input to it. You have that in the repo (function in python that invokes the Perl script)
 
2- script with dependencies
 
If your script has a "lib" with modules it needs to load to work. We upload "lib" and adjust the PERL5LIB accordingly so that Perl will find the modules when it "use"s them. I seem to remember this case is covered in the POC too
 
3- Non system library external dependencies (CPAN)
 
your script depends on some external CPAN module that doesn't need any library from the SO. The plan is to install those dependencies via carton in a machine spawned up from the Lambda AMI (Ithink there is a Docker image too). carton will leave everything installed in "local". Zip up and then from the Lambda function adjust PERL5LIB to include "local/lib/perl". That should make Perl finds the modules it "use"s
 
4- system library dependant CPAN modules
 
Since Lambda won't let you install anything in it's runtime, if a module needs an OS library to work, we need to be creative. It's basically strategy 3 with a bit more stuff, install the system libraries on that machine, install Perl deps with carton, and do the same as in case 3, but this time also zip the needed .so files from the system. In The Lambda function set LD_LIBRARY_PATH accordingly so the so's can be found when Perl needs them.
 
I hope this helps you get a bit closer to your objective. If you want, we can work together on getting some type of streamlined support for executing Perl in Lambda. I haven't been working on it since we don't need it (it was just a POC to get familiar with what happens inside Lambda).
```

It would look that you are a case 3 or 4 . He was a case 3, and told me that:

```
Got my cpanfile built and uploaded the fat library to ec2 instance and it executes.  Woot!
```
But didn't get more feedback (for the moment).

I hope this information helps you get started. Anything that you share back will be great for anyone trying do the same (I think there are more out there :))

**Comment by kittrCZ** [comment]
Hi @pplu, I will take a look on that and will see what we will find!:) Thank you so much for such detailed answer!

**Comment by Palisand** [comment]
@kittrCZ Were you able to successfully host exiftool in Lambda? I keep getting a permission denied error for `exiftool` in spite of setting permissions (755) before uploading. My current function contents are:
```
lib (exiftool dependencies)
exiftool (the perl script)
index.js
```
### Update
Nevermind, I encounter no permission errors when running `perl exiftool` rather than `./exiftool`.

@pplu Thank you for providing this repo!

**Comment by pplu** [comment]
Hi @Palisand 

From where are you getting the exiftool script and dependencies? I think that a permission denied error can also be caused by libraries (.so) files not being found, or maybe by the shebang? Try to capture STDERR (maybe you're missing it)(note: the example in this repo does capture STDERR).

Also you could try to build exiftool in the same environment as Lambda:
Try to download and execute the lambda docker image (see https://github.com/lambci/docker-lambda for more details)
```
docker run -v "$PWD":/var/task lambci/lambda" /bin/bash
```
Then install cpanm (cpan App::cpanminus), and try to do the following: 
```
cd /var/task
cpanm -l local Image::ExifTool
```
That should leave in your current directory a `local/bin/exiftool` and `local/lib/perl5` that should execute in the lambda environment. Try to execute `perl -I local/lib/perl5 local/bin/exiftool` from your index.js

Hope that helps!

**Comment by shubham-gupta-xelp** [comment]
hi @pplu,
how can I install exiftool on lambda?

**Comment by Palisand** [comment]
@shubhamXelpmoc At the top of exiftool's [installation doc](https://www.sno.phy.queensu.ca/~phil/exiftool/install.html), in that red box, it states "ExifTool does not need to be installed to run" and "if you move the Perl 'exiftool' application, you must also move its 'lib' directory to the same location". So you would include exiftool and it's lib directory in your lambda function package.

**Comment by shubham-gupta-xelp** [comment]
@Palisand ,
Where should I add the all files, inside node modules or in the root .

**Comment by Palisand** [comment]
![root](https://www.arborday.org/trees/planting/graphics/bare-root.jpg)

**Comment by shubham-gupta-xelp** [comment]
@Palisand ,
still the same error.

2017-12-24T18:44:51.586Z	86026e02-e8da-11e7-9153-458c414a7d9d	Error: exiftool error:Error: spawn exiftool ENOENT
at ChildProcess.exif.on (/var/task/node_modules/simple-exiftool/lib/index.js:53:25)
at emitOne (events.js:96:13)
at ChildProcess.emit (events.js:188:7)
at Process.ChildProcess._handle.onexit (internal/child_process.js:213:12)
at onErrorNT (internal/child_process.js:367:16)
at _combinedTickCallback (internal/process/next_tick.js:80:11)
at process._tickDomainCallback (internal/process/next_tick.js:128:9)

**Comment by Palisand** [comment]
Did you try running `perl exiftool ...`? Also, I've only tested this on a Node 6 environment.

**Comment by eransteiner** [comment]
To make exiftool work in Lambda:
- Get latest exiftool tag.gz from here: https://sno.phy.queensu.ca/~phil/exiftool/ (at the top)
- Take the exiftool file and lib folder from it and place in your bundle root
- chmod a+x exiftool
- edit node_modules/exiftool/lib/exiftool.js
line 13 change:
var exif = ChildProcess.spawn('exiftool', tags);
to 
var exif = ChildProcess.spawn('/var/task/exiftool', tags);

Then include exiftool and lib in your new ZIP file that you upload to Lambda.

**Comment by pplu** [comment]
Closing this issue. If you want to run Perl code in Lambda, more information is here: https://github.com/pplu/perl-lambda-byor

**Comment by dsreekant211** [comment]
hi @pplu please help me. after build the dockerfile the zip file is not showing .where it will be stored

**Comment by pplu** [comment]
> hi @pplu please help me. after build the dockerfile the zip file is not showing .where it will be stored

Are you following https://github.com/pplu/perl-lambda-byor? Can you please open issue there? Please tell me in which step you are not finding the zip (building the Perl layer? building the dependencies layer? building the zip with your code?)

**Comment by philip-firstorder** [comment]
@Palisand Did you use the preinstalled lambda perl version 5.16 for exiftool or did you install a newer release like 5.28?

Also did you just copy exiftool and lib directory from the official tar.gz without building them on a linux/docker machine?

I am trying to execute perl exiftool -ver from golang and it exits with "error 2". My files are in /var/task (main.go, exiftool, lib/)

Update:
I managed to get it working locally by just adding a precompiled perl Layer from https://metacpan.org/pod/AWS::Lambda, which is extracted to /opt and then I executed `/opt/bin/perl exiftool ...` and it worked.

On production the go lambda function already included perl 5.16, which is enough for exiftool, so adding the precompiled perl layer is not needed. There I just run `perl exiftool ...`

The error 2 that I had was also because I didn‘t recursively zip the /lib folder with -r and it didn‘t have all files.
